File: Behavior/State/Lua/state.lua
-- State 接口定义
State = {}

function State:new()
  local newObj = {}
  self.__index = self
  return setmetatable(newObj, self)
end

function State:handle(context)
end

-- 具体状态：StartState
StartState = State:new()

function StartState:handle(context)
  print("Player is in start state")
  context:setState(StopState:new())
end

-- 具体状态：StopState
StopState = State:new()

function StopState:handle(context)
  print("Player is in stop state")
end

-- Context 类，持有状态
Context = {}

function Context:new()
  local newObj = {state = nil}
  self.__index = self
  return setmetatable(newObj, self)
end

function Context:setState(state)
  self.state = state
end

function Context:request()
  self.state:handle(self)
end


-- main.lua

-- 创建一个 Context 对象并给它设置一个初始状态
local context = Context:new()
context:setState(StartState:new())

-- 应用初始状态的行为
context:request()

-- 状态从 StartState 变为 StopState
context:request()

-- 再次调用，但状态是 StopState，应该没有变化
context:request()

File: Behavior/State/TypeScript/state.ts
// 状态接口，定义了所有具体状态必须实现的方法
interface State {
    handle(context: Context): void;
}

// 上下文类，持有状态的引用并允许状态改变
class Context {
    private state: State;

    constructor(state: State) {
        this.state = state;
    }

    public setState(state: State): void {
        this.state = state;
    }

    public request(): void {
        this.state.handle(this);
    }
}

// 开始状态的具体实现
class StartState implements State {
    public handle(context: Context): void {
        console.log('Player is in start state');
        context.setState(new StopState());
    }
}

// 停止状态的具体实现
class StopState implements State {
    public handle(context: Context): void {
        console.log('Player is in stop state');
        // 可能会设置其他状态，但在这个简单的例子中，我们不做改变
    }
}

// 实例化上下文对象，初始状态设置为 StartState
let context = new Context(new StartState());

// 第一次请求会打印出 "Player is in start state"
// 并将状态改变为 StopState
context.request();

// 第二次请求会打印出 "Player is in stop state"
context.request();

// 第三次请求仍然是 StopState，输出不会改变
context.request();
